### PR TITLE
added spi_get_baudrate() + some consistency changes

### PR DESCRIPTION
**Pull request by Reneg973**
- new function to query the current baudrate
- new helper check whether SPI is busy

+ consistency changes
- return a boolean for functions with a "is; has; can" in name
- return unsigned value for number of bytes read/written
- some const correctness

### CODE REVIEW

**Comment by lurch** [comment]
GitHub says that this has a conflict already?

File: src/rp2_common/hardware_spi/include/hardware/spi.h [hunk 1]
```diff
@@ -197,27 +208,37 @@ static inline void spi_set_slave(spi_inst_t *spi, bool slave) {
...
  *
  * \param spi SPI instance specifier, either \ref spi0 or \ref spi1
- * \return 0 if no space is available to write. Non-zero if a write is possible
+ * \return false if no space is available to write. True if a write is possible
  *
  * \note Although the controllers each have a 8 deep TX FIFO, the current HW implementation can only return 0 or 1
  * rather than the space available.
  */
-static inline size_t spi_is_writable(spi_inst_t *spi) {
+static inline bool spi_is_writable(const spi_inst_t *spi) {
```
[the remark this answers is not in the document]
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:216]: then the function name is wrong, spi_get_fifo_level is what you mean?

File: src/rp2_common/hardware_spi/spi.c [hunk 2]
```diff
@@ -22,11 +22,10 @@ void spi_init(spi_inst_t *spi, uint baudrate) {
     spi_reset(spi);
     spi_unreset(spi);
 
-    (void) spi_set_baudrate(spi, baudrate);
+    spi_set_baudrate(spi, baudrate);
     spi_set_format(spi, 8, SPI_CPOL_0, SPI_CPHA_0, SPI_MSB_FIRST);
     // Always enable DREQ signals -- harmless if DMA is not listening
     hw_set_bits(&spi_get_hw(spi)->dmacr, SPI_SSPDMACR_TXDMAE_BITS | SPI_SSPDMACR_RXDMAE_BITS);
-    spi_set_format(spi, 8, SPI_CPOL_0, SPI_CPHA_0, SPI_MSB_FIRST);
```
Comment (lurch) [inline src/rp2_common/hardware_spi/spi.c:29]: What's the reason for removing this?

Comment (Reneg973) [reply src/rp2_common/hardware_spi/spi.c:29]: it was called 2 lines before

File: src/rp2_common/hardware_spi/include/hardware/spi.h [hunk 3]
```diff
@@ -197,27 +207,37 @@ static inline void spi_set_slave(spi_inst_t *spi, bool slave) {
...
- * \return Non-zero if a read is possible i.e. data is present
+ * \return true if a read is possible i.e. data is present
  *
  * \note Although the controllers each have a 8 deep RX FIFO, the current HW implementation can only return 0 or 1
  * rather than the data available.
  */
-static inline size_t spi_is_readable(spi_inst_t *spi) {
-    return (spi_get_hw(spi)->sr & SPI_SSPSR_RNE_BITS) >> SPI_SSPSR_RNE_LSB;
+static inline bool spi_is_readable(const spi_inst_t *spi) {
+    return (spi_get_hw(spi)->sr & SPI_SSPSR_RNE_BITS);
```
[the remark this answers is not in the document]
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:230]: then we have a compiler bug. bool is defined as 0 or 1 and compiler does this conversion implicitly.

File: src/rp2_common/hardware_spi/include/hardware/spi.h [hunk 4]
```diff
@@ -197,27 +207,37 @@ static inline void spi_set_slave(spi_inst_t *spi, bool slave) {
...
  *
  * \note Although the controllers each have a 8 deep TX FIFO, the current HW implementation can only return 0 or 1
  * rather than the space available.
  */
-static inline size_t spi_is_writable(spi_inst_t *spi) {
+static inline bool spi_is_writable(const spi_inst_t *spi) {
     // PL022 doesn't expose levels directly, so return values are only 0 or 1
-    return (spi_get_hw(spi)->sr & SPI_SSPSR_TNF_BITS) >> SPI_SSPSR_TNF_LSB;
+    return (spi_get_hw(spi)->sr & SPI_SSPSR_TNF_BITS);
 }
```
[the remark this answers is not in the document]
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:218]: I checked with following code:
```
uint8_t yyy = 122;
bool b = yyy;
printf("the value of b is %d", b);

```
=> the value of b is 1

[the remark this answers is not in the document]
Comment (kilograham) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:218]: yeah I thought it was a typedef, but then again that knowledge was probably from long before C99 :-)

File: src/rp2_common/hardware_spi/include/hardware/spi.h [hunk 5]
```diff
@@ -197,27 +207,37 @@ static inline void spi_set_slave(spi_inst_t *spi, bool slave) {
...
  * \param spi SPI instance specifier, either \ref spi0 or \ref spi1
- * \return 0 if no space is available to write. Non-zero if a write is possible
+ * \return false if no space is available to write. True if a write is possible
  *
  * \note Although the controllers each have a 8 deep TX FIFO, the current HW implementation can only return 0 or 1
  * rather than the space available.
  */
-static inline size_t spi_is_writable(spi_inst_t *spi) {
+static inline bool spi_is_writable(const spi_inst_t *spi) {
     // PL022 doesn't expose levels directly, so return values are only 0 or 1
```
Comment (lurch) [inline src/rp2_common/hardware_spi/include/hardware/spi.h:216]: Should this now say "...so return value is a boolean" ?

File: src/rp2_common/hardware_spi/include/hardware/spi.h [hunk 6]
```diff
@@ -132,18 +132,28 @@ void spi_deinit(spi_inst_t *spi);
...
  * \return Number of SPI, 0 or 1.
  */
-static inline uint spi_get_index(spi_inst_t *spi) {
+static inline uint spi_get_index(const spi_inst_t *spi) {
     invalid_params_if(SPI, spi != spi0 && spi != spi1);
     return spi == spi1 ? 1 : 0;
 }
 
-static inline spi_hw_t *spi_get_hw(spi_inst_t *spi) {
+static inline spi_hw_t *spi_get_hw(const spi_inst_t *spi) {
```
[the remark this answers is not in the document]
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:156]: I propose an alternative, add an additional function:
`static inline const spi_hw_t *spi_get_hw_const(const spi_inst_t *spi) {`

Needs some more changes. Agreed?

[the remark this answers is not in the document]
Comment (kilograham) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:156]: we don't do this for any hardware; not least there may be registers for which reading has side effects.

File: src/rp2_common/hardware_spi/include/hardware/spi.h (hunk 5 above)
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:216]: it should

File: src/rp2_common/hardware_spi/include/hardware/spi.h [hunk 7]
```diff
@@ -197,27 +207,34 @@ static inline void spi_set_slave(spi_inst_t *spi, bool slave) {
...
 }
 
 /*! \brief Check whether a read can be done on SPI device
  *  \ingroup hardware_spi
  *
  * \param spi SPI instance specifier, either \ref spi0 or \ref spi1
- * \return Non-zero if a read is possible i.e. data is present
+ * \return true if a read is possible i.e. data is present
  *
  * \note Although the controllers each have a 8 deep RX FIFO, the current HW implementation can only return 0 or 1
```
Comment (lurch) [inline src/rp2_common/hardware_spi/include/hardware/spi.h:223]: Should this "note" be removed too, to match the change you just did for `spi_is_writable`? :slightly_smiling_face:

File: src/rp2_common/hardware_spi/include/hardware/spi.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:156]: Your decision, both ways are possible. Anyway my experience is that const correctness can expose possible issues in code.

E.g. check function  hw_is_claimed(), the function name says it just checks, but in reality it claims, so it does more than expected. One of the rules I try to follow - no surprises. And if hw_is_claimed() modifies something another function (hw_claim_or_assert) is made for, IMO this is a surprise

File: src/rp2_common/hardware_spi/include/hardware/spi.h (hunk 7 above)
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:223]: xD very fussy, but you're right!

**Review by kilograham — CHANGES_REQUESTED** [review]
please rebase onto latest develop (as there is a merge conflict)

File: src/rp2_common/hardware_spi/include/hardware/spi.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h:156]: right, that's ok here, it's just a simple function where const correctness is not of any meaning. I'll remove it

[the remark this answers is not in the document]
Comment (Reneg973) [reply src/rp2_common/hardware_spi/include/hardware/spi.h]: I made a proposal with commit 43da962. Please check whether it's ok or needs a revert

**Review by kilograham — CHANGES_REQUESTED** [review]
try building kitchen_sink ... there are a _bunch_ of compiler warnings

**Comment by kilograham** [comment]
Also, frankly I'm not convinced by switching the return values to size_t... generally we haven't been using size_t stylistically (I know that the params were size_t), however where functions do return a value it tends to have the possibility of an error code.

**Comment by Reneg973** [comment]
> try building kitchen_sink ... there are a _bunch_ of compiler warnings

thanks for the tip. Building this also finds other warnings like unused functions in stdio and a missing delete operator in new_delete.cpp. How do you handle these?

**Comment by Reneg973** [comment]
> however where functions do return a value it tends to have the possibility of an error code.

In general this is fine. Anyway for these functions currently there are 3 arguments against it.
1. the function never returns an error (it is automagically included, if the return is != input len)
2. the documentation doesn't tell anything about possible error code
3. even if the use case may not exist, providing a size_t len which is larger than int will result in a negative value, which is not expected.

Checking for transfer error is still possible by comparing the return value with the input len, but now without a signed/unsigned mismatch comparison.

**Comment by lurch** [comment]
> a missing delete operator in new_delete.cpp

See #304 ?

**Comment by kilograham** [comment]
> > try building kitchen_sink ... there are a _bunch_ of compiler warnings
> 
> thanks for the tip. Building this also finds other warnings like unused functions in stdio and a missing delete operator in new_delete.cpp. How do you handle these?

hmm; what compiler are you using... i don't see those warnings

**Comment by kilograham** [comment]
> > however where functions do return a value it tends to have the possibility of an error code.
> 
> In general this is fine. Anyway for these functions currently there are 3 arguments against it.
> 
>     1. the function never returns an error (it is automagically included, if the return is != input len)
> 
>     2. the documentation doesn't tell anything about possible error code
> 
>     3. even if the use case may not exist, providing a size_t len which is larger than int will result in a negative value, which is not expected.
> 
> 
> Checking for transfer error is still possible by comparing the return value with the input len, but now without a signed/unsigned mismatch comparison.

Yes, i was talking about consistency in the API with other APIs, so this is in general.

1) yes, but doesn't give you the option of return _what_ error
2) sure
3) this is a 32 bit microcontroller; providing a >31 bit value is never going to be valid... this is why size_t wasn't used anywhere really (and when I say anywhere, yes these APIs do use it).

**Comment by Reneg973** [comment]
> hmm; what compiler are you using... i don't see those warnings

PICO_GCC_TRIPLE defaulted to arm-none-eabi
-- The C compiler identification is GNU 10.2.1
-- The CXX compiler identification is GNU 10.2.1

[57/259] Building C object CMakeFiles/kitchen_sink_no_flash.dir/D_/PicoSDK/1.1.2-Git/src/rp2_common/pico_unique_id/unique_id.c.obj
[58/259] Building C object CMakeFiles/kitchen_sink_no_flash.dir/D_/PicoSDK/1.1.2-Git/src/rp2_common/pico_multicore/multicore.c.obj
[59/259] Building C object CMakeFiles/kitchen_sink_no_flash.dir/D_/PicoSDK/1.1.2-Git/src/rp2_common/pico_stdio/stdio.c.obj
FAILED: CMakeFiles/kitchen_sink_no_flash.dir/D_/PicoSDK/1.1.2-Git/src/rp2_common/pico_stdio/stdio.c.obj 
D:\2020-q4-major\bin\arm-none-eabi-gcc.exe -DPARAM_ASSERTIONS_ENABLE_ALL=1 -DPICO_AUDIO_DMA_IRQ=1 -DPICO_BIT_OPS_PICO=1 ... lots_of_includes... pico_stdio/stdio.c
D:/PicoSDK/1.1.2-Git/src/rp2_common/pico_stdio/stdio.c:203:13: error: 'stdio_buffered_printer' defined but not used [-Werror=unused-function]

**Comment by Reneg973** [comment]
> Yes, i was talking about consistency in the API with other APIs, so this is in general.

Ok, so revert?

**Comment by kilograham** [comment]
> > Yes, i was talking about consistency in the API with other APIs, so this is in general.
> 
> Ok, so revert?

yes please, whilst i can see why you might want to, I don't see enough value in changing the API

**Comment by lurch** [comment]
> D:/PicoSDK/1.1.2-Git/src/rp2_common/pico_stdio/stdio.c:203:13: error: 'stdio_buffered_printer' defined but not used [-Werror=unused-function]

Perhaps that function just needs a `#if LIB_PICO_PRINTF_PICO` wrapped around it? https://github.com/raspberrypi/pico-sdk/blob/develop/src/rp2_common/pico_stdio/stdio.c#L219

**Comment by Reneg973** [comment]
ok is reverted. Did you realize that these function in no case return the transferred len? Should I do this with this PR too?

**Comment by kilograham** [comment]
> ok is reverted. Did you realize that these function in no case return the transferred len? Should I do this with this PR too?

do you mean in all cases?

**Comment by Reneg973** [comment]
> > ok is reverted. Did you realize that these function in no case return the transferred len? Should I do this with this PR too?
> 
> do you mean in all cases?

I mean they return the input length and not what was really sent => wrong docu or wrong implementation?

**Comment by lurch** [comment]
> I mean they return the input length and not what was really sent

This may be a stupid question, but for SPI how would those numbers ever be different? :confused:

**Comment by Reneg973** [comment]
> > I mean they return the input length and not what was really sent
> 
> This may be a stupid question, but for SPI how would those numbers ever be different? 😕

Not with SPI itself, of course. But I am sure you heard about bit failures

**Review by kilograham — APPROVED** [review]

**Pull request merged**